### DISCUSSION TITLE
Examples of streaming

### DISCUSSION THREAD

**Discussion by dolencd** [deeplay-io/nice-grpc#249] [Ideas]
First, thanks a lot for this package! It really is nice to you and makes me finally want to use gRPC in a real project 😄 

I did find working with streams a bit difficult, since it's very different than what I (and I'm sure many others) am used to. I've done a bunch of research into async iterables (and learned a lot!) since, but some things stil illude me. In particular, tracking a call that you've made that accepts a stream and send values to it later. Basically what [this example](https://techblog.fexcofts.com/2018/07/20/grpc-nodejs-chat-example/) does when a user enters something in the terminal on the client.

It would be great to have a full example with client and server that use these more novel ways of working with async events. Maybe even a remake of the example linked above.

**Comment by aikoven** [comment]
Thank you for your feedback!

The idiomatic way to produce `AsyncIterable`s for streaming is via async generators. But I agree that it's not always convenient — especially if you want to "push" values to the stream. In this case, you can use [`AsyncSink`](https://github.com/ReactiveX/IxJS/blob/master/src/asynciterable/asyncsink.ts)  from [`IxJS`](https://github.com/ReactiveX/IxJS):

```ts
import {AsyncSink} from "ix/asynciterable";

// AsyncSink is an AsyncIterable that you can push values to
const outgoingMessages = new AsyncSink<Message>();

// start the chat stream
const incomingMessages = chatClient.join(outgoingMessages);

// send message
outgoingMessages.send(...);
```

Hope that helps.

> It would be great to have a full example with client and server that use these more novel ways of working with async events. Maybe even a remake of the example linked above.

Yeah, would be great to have some real-world examples!

**Reply by dolencd** [reply]
Thanks for the suggestion! It's really cool - ends up functioning just like you would expect, but with a completely different data structure underneath 😄
